ranges-lib.tex

```diff
@@ -0,0 +1,797 @@
+%!TEX root = std.tex
+\rSec0[ranges]{Ranges library}
+
+\rSec1[ranges.general]{General}
+
+\pnum
+This Clause describes components for dealing with ranges of elements.
+
+\pnum
+The following subclauses describe
+range and view requirements, and
+components for
+range primitives
+as summarized in Table~\ref{tab:ranges.lib.summary}.
+
+\begin{libsumtab}{Ranges library summary}{tab:ranges.lib.summary}
+  \ref{range.access}        & Range access      & \tcode{<experimental/ranges/range>} \\
+  \ref{range.primitives}    & Range primitives  & \\ \rowsep
+  \ref{ranges.requirements} & Requirements      & \\
+\end{libsumtab}
+
+\rSec1[range.synopsis]{Header \tcode{<experimental/ranges/range>} synopsis}
+
+\indexlibrary{\idxhdr{experimental/ranges/range}}%
+\begin{codeblock}
+#include <experimental/ranges/iterator>
+
+namespace std { namespace experimental { namespace ranges { inline namespace v1 {
+  // \ref{range.access}, range access:
+  namespace {
+    constexpr @\unspec@ begin = @\unspec@;
+    constexpr @\unspec@ end = @\unspec@;
+    constexpr @\unspec@ cbegin = @\unspec@;
+    constexpr @\unspec@ cend = @\unspec@;
+    constexpr @\unspec@ rbegin = @\unspec@;
+    constexpr @\unspec@ rend = @\unspec@;
+    constexpr @\unspec@ crbegin = @\unspec@;
+    constexpr @\unspec@ crend = @\unspec@;
+  }
+
+  // \ref{range.primitives}, range primitives:
+  namespace {
+    constexpr @\unspec@ size = @\unspec@;
+    constexpr @\unspec@ empty = @\unspec@;
+    constexpr @\unspec@ data = @\unspec@;
+    constexpr @\unspec@ cdata = @\unspec@;
+  }
+
+  template <class T>
+  using iterator_t = decltype(ranges::begin(declval<T&>()));
+
+  template <class T>
+  using sentinel_t = decltype(ranges::end(declval<T&>()));
+
+  template <class>
+  constexpr bool disable_sized_range = false;
+
+  template <class T>
+  struct enable_view { };
```

Comment:
WTF `enable_view` is still a class template? Any reason we shouldn't make this a `constexpr bool` variable template like the other user-customizable predicates?